src/Build/Collections/PropertyDictionary.cs

```diff
@@ -529,17 +529,25 @@ internal Dictionary<string, string> ToDictionary()
             }
         }
 
-        internal void Enumerate(Action<string, string> keyValueCallback)
+        internal IEnumerable<(string propertyName, string propertyValue)> Enumerate()
         {
             lock (_properties)
```

Comment:
do you wish to throttle how many critical sections are entered? if so consider type [semaphoreSlim,](https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=net-8.0)

Reply:
I do not intend to change the concurency specific of this code here.
Would we want that - we'd probably need to change underlying structure to be thread safe.

But - not needed in scope of this change and not identified as hot perf issue - so not pursuing now